### PR TITLE
Introduction of RangeIndex

### PR DESCRIPTION
**Pull request by ARF1**
`RangeIndex(1, 10, 2)` is a memory saving alternative to `Index(np.arange(1, 10,2))`: c.f. #939.

This re-implementation is compatible with the current `Index()` api and is a drop-in replacement for `Int64Index()`. It automatically converts to Int64Index() when required by operations.

At present only for a minimum number of operations the type is conserved (e.g. slicing, inner-, left- and right-joins). Most other operations trigger creation of an equivalent Int64Index (or at least an equivalent numpy array) and fall back to its implementation.

This PR also extends the functionality of the `Index()` constructor to allow creation of `RangeIndexes()` with

```
Index(20)
Index(2, 20)
Index(0, 20, 2)
```

in analogy to

```
range(20)
range(2, 20)
range(0, 20, 2)
```

### CODE REVIEW

**Comment by jreback** [comment]
almost all testing is now done in test_index.py
not precluding another file but most / all should be there (in a new section) inheriting from the base tester

**Comment by ARF1** [comment]
@jreback Thanks. I am currently rewriting the test suite of the Int64Index for RangeIndex.

Clearly instantiation will be different. Which of the following cases should be possible?
```
1. RangeIndex(1, 1)
2. RangeIndex(np.nan)
3. RangeIndex([1, 2, 3]) --> I.e. Check if list is monotonically increasing and represent compactly?
4. Index(1) --> RangeIndex(1) ?
```

Should `RangeIndex(5).values == Index(np.arange(5)).values` or something memory saving?

**Comment by shoyer** [comment]
I would not trouble yourself with allowing for non-integer or missing `start`, `step` or `stop`. Also, IMO arguments to `RangeIndex` should be parsed exactly like the built-in `range` (except for the optional `name` keyword argument).

File: pandas/core/range.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+    #   ranges, i.e. the union of RangeIndex(10, 5) and RangeIndex(3, 7) could be
+    #   something like: [RangeIndex(10,5), RangeIndex(3, 7)] and then could just
+    #   iterate over that. But that's for after all of this starts working.
+    # - It would be nice if groupby() accepted an Index
+    # - It might be valuable to cache the values property of a RangeIndex, but
+    #   I'm not totally convinced that's the best strategy (yet!), unless RangeIndex
+    #   could *replace* itself on its parent. (tradeoff between instantiation time and
+    #   ability to gc values when they aren't needed anymore)
+    # TODO: Block setting of start and end
+    def __new__(cls, left, right=None, step=1, name=None):
```
Comment (shoyer) [inline pandas/core/range.py]: I would follow the signature of the builtin range here:
```
class range(stop, *, name=None)
class range(start, stop, step=1, *, name=None)
```

You'll need to parse `args` and `kwargs` manually to make this work.

**Comment by shoyer** [comment]
For implementation, I would strongly recommend keeping track of [builtin `range`](https://docs.python.org/3/library/stdtypes.html#typesseq-range) object (from Python 3 or [a backport](http://python-future.org/reference.html#range)) under the hood. `range` already supports all the bounds checking and slicing logic.

**Comment by ARF1** [comment]
I find that I frequently have to check whether the supplied dtype parameter it equivalent to `np.int64`. Since dtype can be string, etc, my current solution is this:
```
try:
    got_int64_dtype = isinstance(np.dtype(dtype), np.int64)
except TypeError:
    got_int64_dtype = False
```

Is there a testing function already available for use in the idiom `isinstance(some_conv_fn(dtype), np.int64)`?

@shoyer Thanks for the helpful advice and references.

**Comment by jreback** [comment]
this should use virtually the same ``__new__`` as ``Int64Index``. Use ``com.is_integer_dtype``.

**Comment by ARF1** [comment]
Ok, I am slowly making progress: I finally figured out how to do an "analytic" intersection (inner join) between two general, dissimilar `RangeIndex` of the type `range(start, stop, step)`. Of course this results simply in a new `RangeIndex` with a new set of parameters.

Now I am a bit unsure on how to proceed with the union (outer join) of two `RangeIndex`. This can in general not be represented as a single `range(start, stop, step)` but in the worst case requires a combination of three ranges with different parameters.

One could imagine a `CompoundRangeIndex` that keeps a list of multiple `RangeIndex`. 

I was planning to avoid converting `RangeIndex` to `Int64Index` as much as possible but at a certain point operations on a numpy array will probably become faster than running through as list of different `RangeIndex` contained in a `CompoundRangeIndex`...

What does everybody think? Is it worth the effort? Or shall I just convert to a `Int64Index` when a union is required and be done with it?

**Comment by shoyer** [comment]
I would just covert to Int64Index when necessary, possibly even for all union operations on RangeIndex objects (for the sake of consistency).

**Comment by ARF1** [comment]
I am closing this in favor of #9977 to deal with merging issues.

**Pull request closed without merging**